docs/question/server.md

```diff
@@ -134,16 +236,88 @@ def generate(data):
     data["params"]["a"] = random.choice([0.25, 0.5, 1, 2, 4])
 
 def file(data):
-    # We should look at data["filename"], generate the corresponding file,
-    # and return the contents of the file as a string, bytes-like, or file-like object.
-    # We can access data["params"].
-    # As an example, we will generate the "fig.png" figure.
-
-    if data["filename"] == "fig.png":                # check for the appropriate filename
-        plt.plot([0, data["params"]["a"]], [0, 1])   # plot a line with slope "a"
-        buf = io.BytesIO()                           # make a bytes object (a buffer)
-        plt.savefig(buf, format="png")               # save the figure data into the buffer
+    # check for the appropriate filename
+    if data["filename"] == "fig.png":
+        # plot a line with slope "a"
+        plt.plot([0, data["params"]["a"]], [0, 1])
+        # make a bytes object (a buffer)
+        buf = io.BytesIO()
+        # save the figure data into the buffer
+        plt.savefig(buf, format="png")
         return buf
 ```
 
+Note the use of `client_files_question_dynamic_url` in the `question.html` file. Requests for files to this URL will be routed to the `file()` function in `server.py`. The filename requested is stored in `data["filename"]`, and the file contents should be returned from the `file()` function. The URL for the dynamically generated files is set to `{{options.client_files_question_dynamic_url}}` in the HTML, which is a special URL that PrairieLearn uses to route requests to the `file()` function.
```

Comment:
In general, we'd prefer it if people use `<pl-figure>` directly for this; can we update the example above, this paragraph, and the following one to reflect that? I'm happy to keep a reference to `client_files_question_dynamic_url` for advanced users, but that should indeed be an "advanced" alternative to elements that can handle this automatically.

Reply:
> In general, we'd prefer it if people use `<pl-figure>` directly for this

Or `pl-file-download` for downloadable non-images.

> I'm happy to keep a reference to `client_files_question_dynamic_url` for advanced users, but that should indeed be an "advanced" alternative to elements that can handle this automatically.

That reference could point to examples of dynamic files that are neither images nor downloadable files. PDF embeds comes to mind as an example, but I don't know if you want to document that.

Reply:
I updated this section in [6ab669a](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/6ab669a078684c0733dd46b936ece12813b92b53)

Reply:
> Or `pl-file-download` for downloadable non-images.

Yep; I was specifically talking about the example above where we were using `<img src="..." />`, which maps directly to `<pl-figure>`.

> PDF embeds comes to mind as an example, but I don't know if you want to document that.

I think PDF embeds are almost always an anti-pattern compared to just making them available for download.

Reply:
Small wording tweak in [8b77737](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/8b777377f6a58dd85585dd3b2bc2dcff255033b5).

Reply:
Looks good!